crypto/crypto.cpp

```diff
@@ -0,0 +1,169 @@
+// as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+#include <crypto/crypto.hpp>
+#include <openssl/pem.h>
+#include <openssl/err.h>
+
+using namespace bzn;
+
+namespace
+{
+    const std::string PEM_PREFIX = "-----BEGIN PUBLIC KEY-----\n";
+    const std::string PEM_SUFFIX = "\n-----END PUBLIC KEY-----\n";
+}
+
+crypto::crypto(std::shared_ptr<bzn::options_base> options)
+        : options(std::move(options))
+{
+}
+
+void
+crypto::start()
+{
+    if(this->options->get_simple_options().get<bool>(bzn::option_names::CRYPTO_ENABLED_OUTGOING))
+    {
+        this->load_private_key();
+    }
+}
+
+bool
+crypto::verify(const wrapped_bzn_msg& msg)
+{
+    BIO* bio = NULL;
+    EC_KEY* pubkey = NULL;
+    EVP_PKEY* key = NULL;
+    EVP_MD_CTX* context = NULL;
+
+    // This is admittedly a weird structure, but it seems the cleanest way to deal with the combined manual
+    // memory management and error handling
+
+    bool result =
+            // Reconstruct the PEM file in memory (this is awkward, but it avoids dealing with EC specifics)
+            (bio = BIO_new(BIO_s_mem()))
+            && (0 < BIO_write(bio, PEM_PREFIX.c_str(), PEM_PREFIX.length()))
+            && (0 < BIO_write(bio, msg.sender().c_str(), msg.sender().length()))
+            && (0 < BIO_write(bio, PEM_SUFFIX.c_str(), PEM_SUFFIX.length()))
+
+            // Parse the PEM string to get the public key the message is allegedly from
+            && (pubkey = PEM_read_bio_EC_PUBKEY(bio, NULL, NULL, NULL))
+            && (1 == EC_KEY_check_key(pubkey))
+            && (key = EVP_PKEY_new())
+            && (1 == EVP_PKEY_set1_EC_KEY(key, pubkey))
+
+            // Perform the signature validation
+            && (context = EVP_MD_CTX_new())
+            && (1 == EVP_DigestVerifyInit(context, NULL, EVP_sha512(), NULL, key))
+            && (1 == EVP_DigestVerifyUpdate(context, msg.payload().c_str(), msg.payload().length()))
+            && (1 == EVP_DigestVerifyFinal(context, reinterpret_cast<const unsigned char*>(msg.signature().c_str()), msg.signature().length()));
+
+    if (context) EVP_MD_CTX_free(context);
+    if (pubkey) OPENSSL_free(pubkey);
+    if (key) OPENSSL_free(key);
+    if (bio) OPENSSL_free(bio);
+
+    /* Any errors here can be attributed to a bad (potentially malicious) incoming message, and we we should not
+     * pollute our own logs with them (but we still have to clear the error state)
+     */
+    ERR_clear_error();
+
+
+    return result;
+}
+
+bool
+crypto::sign(wrapped_bzn_msg& msg)
```

Comment:
I like to add bzn:: at least in the function parameters

Reply:
wrapped_bzn_msg isn't actually in bzn; it's from protobuf